controllers/atlasmigration_controller.go

```diff
@@ -16,39 +16,310 @@ limitations under the License.
+		return ctrl.Result{}, nil
+	}
+
+	// Reconcile given resource
+	am.Status, err = r.reconcile(ctx, am)
+	if err != nil {
+		// Verify permanent errors to avoid infinite reconciliation loop
+		if isPernamentError(err) {
+			return result(err)
+		}
+
+		return result(transient(err))
+	}
+
+	log.Info(
+		"reconciled successfully",
+		"lastAppliedVersion", am.Status.LastAppliedVersion,
+		"lastApplied", time.Unix(am.Status.LastApplied, 0).String(),
+		"lastDeploymentUrl", am.Status.LastDeploymentURL,
+	)
 	return ctrl.Result{}, nil
 }
 
+// Reconcile the given AtlasMigration resource.
+func (r *AtlasMigrationReconciler) reconcile(
+	ctx context.Context,
+	am dbv1alpha1.AtlasMigration) (dbv1alpha1.AtlasMigrationStatus, error) {
+
+	// Build HCL template data
+	hclTmplData, cleanUpDir, err := r.buildHCLTmplData(ctx, am)
+	if err != nil {
+		return dbv1alpha1.AtlasMigrationStatus{}, err
+	}
+	defer cleanUpDir()
+
+	// Create atlas.hcl from template data
+	file, cleanUpFile, err := executeHCLTemplate("migrateconf.tmpl", hclTmplData)
+	if err != nil {
+		return dbv1alpha1.AtlasMigrationStatus{}, err
+	}
+	defer cleanUpFile()
+
+	// Execute Atlas CLI migrate command
+	report, err := r.CLI.Apply(ctx, &atlas.ApplyParams{ConfigURL: file})
+	if err != nil {
+		return dbv1alpha1.AtlasMigrationStatus{}, err
+	}
+
+	return dbv1alpha1.AtlasMigrationStatus{
+		LastApplied:        report.End.Unix(),
+		LastAppliedVersion: report.Current,
+	}, nil
+}
+
+func (r *AtlasMigrationReconciler) buildHCLTmplData(
+	ctx context.Context,
+	am dbv1alpha1.AtlasMigration) (HCLTmplData, func() error, error) {
+	tmplData := HCLTmplData{}
+	err := error(nil)
+
+	// Get database connection string
+	tmplData.URL = am.Spec.URL
+	if tmplData.URL == "" {
+		tmplData.URL, err = r.getSecretValue(ctx, am.Namespace, *am.Spec.URLFrom.SecretKeyRef)
+		if err != nil {
+			return tmplData, nil, err
+		}
+	}
+
+	// Get temporary directory
+	cleanUpDir := func() error { return nil }
+	if am.Spec.Dir.ConfigMapRef != "" {
+		tmplData.Migration.Dir, cleanUpDir, err = r.createTmpDir(ctx, am.Namespace, am.Spec.Dir)
+		if err != nil {
+			return tmplData, nil, err
+		}
+	}
+
+	// Get Atlas Cloud Token from secret
+	if am.Spec.Cloud.TokenFrom.SecretKeyRef != nil {
+		tmplData.Cloud.Token, err = r.getSecretValue(ctx, am.Namespace, *am.Spec.Cloud.TokenFrom.SecretKeyRef)
+		if err != nil {
+			return tmplData, nil, err
+		}
+	}
+
+	// Mapping cloud fields if token is present
+	if tmplData.Cloud.Token != "" {
+		tmplData.Cloud.URL = am.Spec.Cloud.URL
+		tmplData.Cloud.Project = am.Spec.Cloud.Project
+		tmplData.Data.RemoteDir.Name = am.Spec.Dir.Remote.Name
+		tmplData.Data.RemoteDir.Tag = am.Spec.Dir.Remote.Tag
+	}
+
+	return tmplData, cleanUpDir, nil
+}
+
+// Get the value of the given secret key selector.
+func (r *AtlasMigrationReconciler) getSecretValue(
+	ctx context.Context,
+	ns string,
+	selector corev1.SecretKeySelector) (string, error) {
```

Comment:
No break line between args.
```suggestion
func (r *AtlasMigrationReconciler) getSecretValue(ctx context.Context, ns string, selector corev1.SecretKeySelector) (string, error) {
```

Reply:
I will fix it. By the way, we need a tool to automatically format the code styling.

Reply:
the reason why I put line breaks between args because when you open the file in IDE, it will look something like this
<img width="995" alt="Screenshot 2023-05-25 at 19 32 46" src="https://github.com/ariga/atlas-operator/assets/16095902/e975d04b-9072-410d-a651-63a881d58532">